AirCasting/ABConnector/SDCardAirBeamServices.swift

```diff
@@ -0,0 +1,104 @@
+// Created by Lunar on 17/11/2021.
+//
+
+import CoreBluetooth
+import CoreMIDI
+
+enum SDCardSessionType: CaseIterable {
+    case mobile, fixed, cellular
+}
+
+struct SDCardDataChunk {
+    let payload: String
+    let sessionType: SDCardSessionType
+}
+
+protocol SDCardAirBeamServices {
+    func downloadData(from peripheral: CBPeripheral, progress: @escaping (SDCardDataChunk) -> Void, completion: @escaping (Result<Void, Error>) -> Void)
+}
+
+class BluetoothSDCardAirBeamServices: SDCardAirBeamServices {
+    private let singleChunkMeasurementsCount = 4
+    // has notifications about measurements count in particular csv file on SD card
+    private let DOWNLOAD_META_DATA_FROM_SD_CARD_CHARACTERISTIC_UUID = CBUUID(string:"0000ffde-0000-1000-8000-00805f9b34fb")
+    
+    // has notifications for reading measurements stored in csv files on SD card
+    private let DOWNLOAD_FROM_SD_CARD_CHARACTERISTIC_UUID = CBUUID(string:"0000ffdf-0000-1000-8000-00805f9b34fb")
+    
+    private let bluetoothManager: BluetoothManager
+    
+    private var currentSessionType: SDCardSessionType?
+    private var currentSessionTypeReceived: Int = 0
+    private var currentSessionTypeExpected: Int = 0
+    
+    private var dataCharacteristicObserver: AnyHashable?
+    private var metadataCharacteristicObserver: AnyHashable?
+    
+    init(bluetoothManager: BluetoothManager) {
+        self.bluetoothManager = bluetoothManager
+    }
+    
+    func downloadData(from peripheral: CBPeripheral, progress: @escaping (SDCardDataChunk) -> Void, completion: @escaping (Result<Void, Error>) -> Void) {
+        metadataCharacteristicObserver = bluetoothManager.subscribeToCharacteristic(DOWNLOAD_META_DATA_FROM_SD_CARD_CHARACTERISTIC_UUID) { result in
+            switch result {
+            case .success(let data):
+                guard let data = data, let payload = String(data: data, encoding: .utf8) else {
+                    return
+                }
+                self.currentSessionType = self.currentSessionType.next
+                
+                Log.info(payload)
+                // Payload format is ` Some string: ${number_of_entries_expected} `
+                if payload == "SD_SYNC_FINISH" {
+                    completion(.success(()))
```

Comment:
those 3 lines are repeated in here and in failure and are responsible for exactly the same logical thing, so a good idea would be to put them in a separate function and only call it here. For example
```Swift
finishSynchronization { completion(.success()) } // and .failure(error) for the failed case
```